smarts/core/plan.py

```diff
@@ -335,45 +335,41 @@ def create_route(self, mission: Mission):
 
         assert isinstance(self._mission.goal, PositionalGoal)
 
-        start_lane = self._road_map.nearest_lane(
+        start_lanes = self._road_map.nearest_lanes(
             self._mission.start.point,
-            include_junctions=False,
+            include_junctions=True,
         )
-
-        if not start_lane:
-            # it's possible that the Mission's start point wasn't explicitly
-            # specified by a user, but rather determined during the scenario run
-            # from the current position of a vehicle, in which case it may be
-            # in a junction.  But we only allow this if the previous query fails.
-            start_lane = self._road_map.nearest_lane(
-                self._mission.start.point,
-                include_junctions=True,
-            )
-        if start_lane is None:
+        if not start_lanes:
             self._mission = Mission.endless_mission(Pose.origin())
-            raise PlanningError("Cannot find start lane. Route must start in a lane.")
-        start_road = start_lane.road
+            raise PlanningError("Starting lane not found. Route must start in a lane.")
+
+        via_roads = [self._road_map.road_by_id(via) for via in self._mission.route_vias]
 
         end_lane = self._road_map.nearest_lane(
             self._mission.goal.position,
             include_junctions=False,
         )
         assert end_lane is not None, "route must end in a lane"
-        end_road = end_lane.road
-
-        via_roads = [self._road_map.road_by_id(via) for via in self._mission.route_vias]
 
-        self._route = self._road_map.generate_routes(
-            start_road, end_road, via_roads, 1
-        )[0]
+        # When an agent is in an intersection, the `nearest_lanes` method might
+        # not return the correct road as the first choice. Hence, nearest
+        # starting lanes are tried in sequence until a route is found or until
+        # all nearby starting lane options are exhausted.
+        for start_lane, _ in start_lanes:
+            self._route = self._road_map.generate_routes(
+                start_lane.road, end_lane.road, via_roads, 1
+            )[0]
+            if self._route.road_length > 0:
+                break
```

Comment:
This breaks on the first route. Is seems to go counter to the comment above.

Reply:
The `nearest_lane()` method returns nearby `start_lanes` sorted in increasing distance from the given point. Previously, we tried to find a route using the first `start_lane` returned, which might not yield a route if the `start_lane` was incorrect. Hence, now we try each `start_lane` in a loop until we find a legal route and immediately break upon finding one with `_route.road_length > 0` . We do not wait to compute routes for all the `start_lanes` and pick the shortest overall route, because the further `start_lanes` are away from the given point the higher the chances of it also being an incorrect starting road.

Reply:
OK, I just note that there is the following potential problem.
![selection_problem](https://user-images.githubusercontent.com/8764693/208003933-3f660730-f114-449f-8914-e013baad4950.jpg)